### PR TITLE
WIP: node: implement lazy frame r/w and relaying

### PR DESCRIPTION
**Pull request by jcorbin**
r @Raynos @kriskowal

### CODE REVIEW

**Comment by jcorbin** [comment]
Early share, not done yet.

This only gives a disappointing ~25% boost on the relay benchmark locally for me, perhaps:
- gc churn is the more important long poll; so switching to object pools may be a Big Deal
- maybe my code's not as active as I think it is, I haven't yet fully verified that the relay demo is actually using lazy r/w all the time

Items still TODO:
- [ ] more lazy r/w coverage (e.g. headers, arg1, etc)
- [ ] implement timeout support for `LazyRelayInReq` and `LazyRelayOutReq`
- [ ] implement circuit breaking and stats hook points for same
- [ ] consider alternate approach using `bufrw.buf2` directly as the read machine chunkRW, and then all lazy manipulations are just buffer taking functions only (eliminates `LazyFrame`)

**Comment by Raynos** [comment]
Please also:

 - [ ] code coverage on relay_request and relay_handler before & after
 - [ ] npm link tchannel into rt/autobahn and run that test suite.

I can jump in & help with code coverage

Before we deploy this change:

 - [ ] get stats working
 - [ ] get peer state integration working

We really need to do a good run at improving our tests before we switch over with the removal of InRequest and OutRequest

File: node/v2/index.js [hunk 1]
```diff
@@ -27,6 +27,7 @@ module.exports.Types = Types;
 
 var RetryFlags = require('../retry-flags.js');
 var Frame = require('./frame');
+var LazyFrame = require('./frame');
```
Comment (Raynos) [inline node/v2/index.js]: Needs to be `require('./lazy_frame')`

When trying to run benchmarks I get

```
failed to setup clients { [TchannelProtocolReadFailedError: tchannel read failure: short read, 106 byte left over after consuming 8]
  type: 'tchannel.protocol.read-failed',
  message: 'tchannel read failure: short read, 106 byte left over after consuming 8',
  remoteName: null,
  localName: null,
  name: 'TchannelProtocolReadFailedError',
  frameId: undefined,
  causeMessage: 'short read, 106 byte left over after consuming 8',
  origMessage: 'short read, 106 byte left over after consuming 8',
  fullType: 'tchannel.protocol.read-failed~!~bufrw.short-read' }
```

Comment (jcorbin) [inline node/v2/index.js]: oh man... I wonder if that means that I'm not even doing anything different then ... lol

Comment (jcorbin) [inline node/v2/index.js]: in other news... with the newly lazy frame being added but dead path, I'm now amazed that there was any performance difference from master

**Comment by jcorbin** [comment]
not yet working, but closer, pausing for now

File: node/v2/index.js (hunk 1 above)
Comment (Raynos) [inline node/v2/index.js]: Did you do a comparison with master or just eyeball it ?

Might just be a difference in machines.

**Comment by jcorbin** [comment]
Now showing nearly 2x speedup on the relay benchmark.

More importantly, top now shows the multi_bench server and the relay server jockeying for top cpu use, rather than a perma-pegged relay.

```
% readlink benchmarks/relay-master.json
relay-17a9997.json
% readlink benchmarks/relay-lazy_relay.json
relay-a2eb211.json
% node benchmarks/compare.js benchmarks/relay-master.json benchmarks/relay-lazy_relay.json
GET 16KiB, 1000/0  rate: hi-diff:   3273.24 (106.1%)
GET 16KiB, 10000/0 rate: hi-diff:   2772.42 ( 96.7%)
GET 16KiB, 20000/0 rate: hi-diff:   2690.54 ( 93.5%)
GET 4B, 1000/0     rate: hi-diff:   2026.73 ( 41.8%)
GET 4B, 10000/0    rate: hi-diff:   2079.36 ( 43.3%)
GET 4B, 20000/0    rate: hi-diff:   4242.99 ( 87.6%)
GET 4KiB, 1000/0   rate: hi-diff:   2650.40 ( 65.7%)
GET 4KiB, 10000/0  rate: hi-diff:   1466.30 ( 34.1%)
GET 4KiB, 20000/0  rate: hi-diff:   3243.42 ( 81.3%)
PING, 1000/0       rate: hi-diff:   1348.47 ( 24.1%)
PING, 10000/0      rate: hi-diff:   2302.08 ( 47.3%)
PING, 20000/0      rate: hi-diff:   3344.15 ( 66.3%)
SET 16KiB, 1000/0  rate: hi-diff:   2876.63 ( 79.8%)
SET 16KiB, 10000/0 rate: hi-diff:   2999.34 (114.8%)
SET 16KiB, 20000/0 rate: hi-diff:   2753.10 ( 94.9%)
SET 4B, 1000/0     rate: hi-diff:   1855.85 ( 37.7%)
SET 4B, 10000/0    rate: hi-diff:   1977.12 ( 39.5%)
SET 4B, 20000/0    rate: hi-diff:   3712.73 ( 76.7%)
SET 4KiB, 1000/0   rate: hi-diff:   2506.68 ( 64.4%)
SET 4KiB, 10000/0  rate: hi-diff:   3485.36 ( 92.6%)
SET 4KiB, 20000/0  rate: hi-diff:   3635.29 (102.4%)
```

**Comment by Raynos** [comment]
Nice :)

**Comment by ShanniLi** [comment]
:+1:

**Comment by breerly** [comment]
:100:

**Comment by Raynos** [comment]
Rebased.

**Pull request closed without merging**